### PR TITLE
[RFC] Add support for compact protocol

### PR DESCRIPTION
**Pull request by bgamari**
Thanks for the great library. This is the current state of my ill-fated attempt at adding support for the compact protocol. I say ill-fated because it unfortunately seems impossible to correctly implement this protocol within the confines of the existing interface.

While this patch implements most of the protocol correctly, it unfortunately falls short in handling empty maps and other containers. In this case the compact protocol omits any type information, which means we are unable to provide an unambiguous `TType` for the resulting `Value`. Unfortunately it's really not clear how to fix this while maintaining the nice semantics of the library (e.g. allowing access to a `Value`'s `TType`). In this sense it appears that the compact protocol sacrifices the fully self-describing property that the original binary protocol preserved.

I can see two ways forward, neither very appealing,
 1. Require that type information is provided during deserialization
 2. Add a `TAny` type, meaning that the type of a term depends upon its value (since we would need to assign an empty list the element type `TAny` whereas a non-empty type would have a proper type.
 
Perhaps you have some ideas on how to proceed?

### CODE REVIEW

**Comment by abhinav** [comment]
Hi, thanks for implementing this! I'll give the code a look tonight.

I didn't realize that the compact protocol loses that information. Makes sense, though. I think we can get away with an approach like `TAny`, but calling it `TStop` instead (this appears to be closer to what the official implementations do).

We can introduce a new `TType`, `TStop`, which is invalid if passed to serialize/deserialize. There will be no `Value` constructors that build a value with type `TStop`. When parsing, if the map size is 0, an empty list of `MapItem TStop TStop` will be returned. Something like,

```haskell
if count == 0
  then return $ VMap (FL.empty :: FL.FoldList (MapItem TStop TStop))
  else parseStuff
```

This will type check just fine because `VMap` is just `Value TMap` and the inner types aren't exposed.

The `Pinchable` instances for maps will simply have to special case `TStop` as "don't look at the value", it's empty. Although, I suspect that the special casing won't be necessary because the logic to [`unpinch` a map](https://github.com/abhinav/pinch/blob/master/src/Pinch/Internal/Pinchable.hs#L223) doesn't look at the `TType`s associated with the map, it looks at the `TType`s of the individual elements in the map. So for an empty map, it won't care that the `TType` doesn't match the type it's expecting to parse, and it'll just work.

Does that sound like a good approach?

**Comment by bgamari** [comment]
Abhinav Gupta <notifications@github.com> writes:

> Hi, thanks for implementing this! I'll give the code a look tonight.
>
Sure!
> I didn't realize that the compact protocol loses that information.
> Makes sense, though. I think we can get away with an approach like
> `TAny`, but calling it `TStop` instead (this appears to be closer to
> what the official implementations do).
>
It turns out that modifying the type system isn't strictly necessary.
Indeed all that is necessary is to add a constructor to Value, `VNullMap
:: Value TMap`. This is what the patch now does; let me know what you
think.

Otherwise all of the tests currently pass so I don't expect there will
be any further challenges.

Cheers,

- Ben

File: src/Pinch/Internal/Value.hs [hunk 1]
```diff
@@ -70,6 +70,7 @@ data Value a where
 
     VMap  :: forall k v. (IsTType k, IsTType v)
           => !(FoldList (MapItem k v)) -> Value TMap
+    VNullMap :: Value TMap
```
Comment (abhinav) [inline src/Pinch/Internal/Value.hs]: You'll need to handle `VNullMap` in the `Pinchable` instances for maps. Currently they only expect a `VMap`.

File: pinch.cabal [hunk 2]
```diff
@@ -70,6 +70,7 @@ library
         Pinch.Internal.Value
         Pinch.Protocol
         Pinch.Protocol.Binary
+        Pinch.Protocol.Compact
```
Comment (abhinav) [inline pinch.cabal]: Hey. This project makes use of [hpack](https://github.com/sol/hpack). You should add this entry to package.yaml instead.

File: src/Pinch/Internal/Value.hs [hunk 3]
```diff
@@ -191,6 +194,7 @@ instance Hashable (Value a) where
       VInt64  x -> s `hashWithSalt` (6 :: Int) `hashWithSalt` x
       VList   x -> s `hashWithSalt` (7 :: Int) `hashWithSalt` x
       VMap    x -> s `hashWithSalt` (8 :: Int) `hashWithSalt` x
+      VNullMap  -> s `hashWithSalt` (8 :: Int)
```
Comment (abhinav) [inline src/Pinch/Internal/Value.hs]: We should use a different salt value for each constructor. I would make this 11.

File: src/Pinch/Protocol/Compact.hs [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+import qualified Pinch.Internal.Parser   as P
+
+
+-- | Provides an implementation of the Thrift Compact Protocol.
+compactProtocol :: Protocol
+compactProtocol = Protocol
+    { serializeValue     = binarySerialize
+    , deserializeValue'  = binaryDeserialize ttype
+    , serializeMessage   = binarySerializeMessage
+    , deserializeMessage = binaryDeserializeMessage
```
Comment (abhinav) [inline src/Pinch/Protocol/Compact.hs]: Just a nit: These methods were called `binary{Serialize,Deserialize}*` because they were implementing the binary protocol. Here, you would want to call them `compactSerialize`, `compactDeserialize`, etc. or just `serialize`, `deserialize`.

File: src/Pinch/Internal/Value.hs [hunk 5]
```diff
@@ -89,6 +90,7 @@ instance Show (Value a) where
         go xs i (SomeValue val) = (show i ++ ": " ++ show val):xs
 
     show (VMap x) = show x
+    show VNullMap = "[]"
```
Comment (abhinav) [inline src/Pinch/Internal/Value.hs]: Can you also update the `NFData` instance please?

File: src/Pinch/Protocol/Compact.hs [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+#if __GLASGOW_HASKELL__ < 709
+import Control.Applicative
+#endif
+
+import Control.Monad
+import Data.List           (sortBy)
+import Data.Ord            (comparing)
+import Data.Bits           hiding (shift)
+import Data.ByteString     (ByteString)
+import Data.Foldable       (null)
```
Comment (abhinav) [inline src/Pinch/Protocol/Compact.hs]: This isn't available in GHC 7.8. I think you can implement this in FoldList:

    null (FoldList f) = f (\_ _ -> False) True

File: src/Pinch/Internal/Value.hs (hunk 1 above)
Comment (abhinav) [inline src/Pinch/Internal/Value.hs]: Just checked. This is just a matter of adding this to `unpinchMap`:

```haskell
unpinchMap _ mapEmpty VNullMap = return mapEmpty
```

Comment (abhinav) [inline src/Pinch/Internal/Value.hs]: It might be worth adding a test that round trips an actual empty map (not a `Value`) to compact and back.

File: src/Pinch/Internal/Value.hs (hunk 3 above)
Comment (bgamari) [inline src/Pinch/Internal/Value.hs]: I'm not so certain about this; the problem is that if you distinguish these cases then you lose the ability to round-trip values. This would be extremely surprising to the user. Really, `VNullMap` is an implementation detail which should be hid as much as possible. Really the fact that it's user-visible at all is IMHO problematic.

File: src/Pinch/Protocol/Compact.hs (hunk 4 above)
Comment (bgamari) [inline src/Pinch/Protocol/Compact.hs]: Alright, I'll take care of this.

File: src/Pinch/Internal/Value.hs (hunk 3 above)
Comment (abhinav) [inline src/Pinch/Internal/Value.hs]: That makes sense. Disregard this comment.

File: src/Pinch/Protocol/Compact.hs [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+    { serializeValue     = binarySerialize
+    , deserializeValue'  = binaryDeserialize ttype
+    , serializeMessage   = binarySerializeMessage
+    , deserializeMessage = binaryDeserializeMessage
+    }
+
+------------------------------------------------------------------------------
+
+bword8 :: Word8 -> Builder
+bword8 = BB.int8 . fromIntegral
```
Comment (abhinav) [inline src/Pinch/Protocol/Compact.hs]: This is leftover from when you added `word8` to `Builder`. I would replace uses of this with `BB.word8`.

File: src/Pinch/Protocol/Compact.hs [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+    pid <- P.word8
+    when (pid /= protocolId) $ fail "Invalid protocol ID"
+    w <- P.word8
+    let ver = w .&. 0x1f
+    when (ver /= version) $ fail $ "Unsupported version: " ++ show ver
+    let code = w `shiftR` 5
+    msgId <- parseVarint
+    msgName <- TE.decodeUtf8 <$> (parseVarint >>= P.take . fromIntegral)
+    payload <- binaryParser ttype
+    return Message { messageType = case fromMessageCode code of
```
Comment (abhinav) [inline src/Pinch/Protocol/Compact.hs]: You should make unknown message type a parser failure.

```haskell
mtype <- case fromMessageCode code of
  Nothing -> fail $ "Unknown message type: " ++ show code
  Just mt -> return mt
```

File: src/Pinch/Protocol/Compact.hs [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+parseInt32 :: Parser (Value TInt32)
+parseInt32 = VInt32 . fromIntegral . zigZagToInt <$> parseVarint
+
+parseInt64 :: Parser (Value TInt64)
+parseInt64 = VInt64 . fromIntegral . zigZagToInt <$> parseVarint
+
+parseBinary :: Parser (Value TBinary)
+parseBinary = do
+    n <- parseVarint
+    when (n < 0) $ fail "parseBinary: invalid length"
```
Comment (abhinav) [inline src/Pinch/Protocol/Compact.hs]: Can you include the length in the error message? Should help with debugging if this is ever raised.

File: src/Pinch/Protocol/Compact.hs [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+      0 -> return VNullMap
+      _ -> do
+          tys <- P.word8
+          ktype' <- getCType (tys `shiftR` 4)
+          vtype' <- getCType (tys .&. 0x0f)
+
+          case (ktype', vtype') of
+            (SomeCType kctype, SomeCType vctype) -> do
+              ktype <- pure $ cTypeToTType kctype
+              vtype <- pure $ cTypeToTType vctype
```
Comment (abhinav) [inline src/Pinch/Protocol/Compact.hs]: Is there a reason you chose to do `x <- pure foo` instead of `let x = foo`?

File: src/Pinch/Protocol/Compact.hs [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+  TInt64   -> serializeInt64
+  TStruct  -> serializeStruct
+  TList    -> serializeList
+  TMap     -> serializeMap
+  TSet     -> serializeSet
+{-# INLINE binarySerialize #-}
+
+serializeBinary :: Value TBinary -> Builder
+serializeBinary (VBinary x) =
+    serializeVarint (fromIntegral $ B.length x) <> BB.byteString x
```
Comment (abhinav) [inline src/Pinch/Protocol/Compact.hs]: This can just be `string x`

**Comment by abhinav** [comment]
Hey. I'm done reviewing this. This looks great. I mostly had stylistic concerns. 

Also, the build seems to be broken on all versions of GHC. For pre-7.10, it's because of:

```
src/Pinch/Internal/Value.hs:168:45:
    Not in scope: type constructor or class ‘Foldable’
    Perhaps you meant ‘F.Foldable’ (imported from Data.Foldable)
```

For 7.10 and above, it's because of:

```
tests/Pinch/Protocol/CompactSpec.hs:21:8:
    Could not find module ‘Pinch.Protocol.Compact’
    it is a hidden module in the package ‘pinch-0.2.0.1@pinch_CxCWmmHvJQVLA3tJTm4NVM’
```

I suspect this one is happening because the new module is not listed in package.yaml as an exposed module and `hpack` overwrites your change to the cabal file, defaulting the new module to `other-modules`.

File: pinch.cabal (hunk 2 above)
Comment (abhinav) [inline pinch.cabal]: Addressed in bgamari#1

File: src/Pinch/Internal/Value.hs (hunk 1 above)
Comment (abhinav) [inline src/Pinch/Internal/Value.hs]: Addressed in https://github.com/bgamari/pinch/pull/2

File: src/Pinch/Internal/Value.hs (hunk 5 above)
Comment (abhinav) [inline src/Pinch/Internal/Value.hs]: Addressed in https://github.com/bgamari/pinch/pull/3

**Comment by abhinav** [comment]
I addressed a couple of minor issues (including the >=7.10 failures) and created pull requests against your PR branch.

- https://github.com/bgamari/pinch/pull/1
- https://github.com/bgamari/pinch/pull/2
- https://github.com/bgamari/pinch/pull/3

You can merge them in with the changes you're already making. Let me know if you need help getting this finished. 

Abhinav

File: src/Pinch/Protocol/Compact.hs (hunk 7 above)
Comment (abhinav) [inline src/Pinch/Protocol/Compact.hs]: My bad. You didn't add word8 to Builder.

File: src/Pinch/Protocol/Compact.hs (hunk 6 above)
Comment (abhinav) [inline src/Pinch/Protocol/Compact.hs]: Also, it turns out that the null check on FoldList is really expensive because it'll have to iterate through the whole collection. Instead, we can use the sizes which we get anyway when we are serializing the objects.

**Comment by abhinav** [comment]
Merged. Thanks!

**Comment by abhinav** [comment]
This is out now in pinch-0.3.0.0.

**Comment by bgamari** [comment]
Abhinav Gupta <notifications@github.com> writes:

> This is out now in pinch-0.3.0.0.
>
Thanks!
